### PR TITLE
Add Reactive One-Time Token Login support

### PR DESCRIPTION
**Pull request by CrazyParanoid**
Closes gh-15699

### CODE REVIEW

File: config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java [hunk 1]
```diff
@@ -5686,4 +5735,278 @@ private AnonymousSpec() {
...
+			Assert.notNull(oneTimeTokenService, "oneTimeTokenService cannot be null");
+			this.oneTimeTokenService = oneTimeTokenService;
+			return this;
+		}
+
+		ReactiveOneTimeTokenService getOneTimeTokenService() {
+			if (this.oneTimeTokenService == null) {
+				this.oneTimeTokenService = new InMemoryReactiveOneTimeTokenService();
+			}
+			return this.oneTimeTokenService;
```
Comment (marcusdacoregio) [inline config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java:5916]: You can probably initialize the `oneTimeTokenService` field with `InMemoryReactiveOneTimeTokenService` and remove this method

File: config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java [hunk 2]
```diff
@@ -5686,4 +5735,278 @@ private AnonymousSpec() {
...
+			Assert.notNull(authenticationConverter, "authenticationConverter cannot be null");
+			this.authenticationConverter = authenticationConverter;
+			return this;
+		}
+
+		ServerAuthenticationConverter getAuthenticationConverter() {
+			if (this.authenticationConverter == null) {
+				return new ServerOneTimeTokenAuthenticationConverter();
+			}
+			return this.authenticationConverter;
```
Comment (marcusdacoregio) [inline config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java:5935]: You can probably initialize the `authenticationConverter` field with `ServerOneTimeTokenAuthenticationConverter` and remove this method

File: web/src/main/java/org/springframework/security/web/server/authentication/ott/GenerateOneTimeTokenWebFilter.java [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+public final class GenerateOneTimeTokenWebFilter implements WebFilter {
+
+	private final ReactiveOneTimeTokenService oneTimeTokenService;
+
+	private ServerWebExchangeMatcher matcher = ServerWebExchangeMatchers.pathMatchers(HttpMethod.POST, "/ott/generate");
+
+	private ServerGeneratedOneTimeTokenHandler generatedOneTimeTokenHandler = new ServerRedirectGeneratedOneTimeTokenHandler(
+			"/login/ott");
+
+	public GenerateOneTimeTokenWebFilter(ReactiveOneTimeTokenService oneTimeTokenService) {
```
Comment (marcusdacoregio) [inline web/src/main/java/org/springframework/security/web/server/authentication/ott/GenerateOneTimeTokenWebFilter.java:49]: Can you please require the `ServerGeneratedOneTimeTokenHandler` in the constructor? Like https://github.com/spring-projects/spring-security/commit/a88a7744ed7fe0116c4c9f713d59a8c65c389595

**Review by marcusdacoregio — CHANGES_REQUESTED** [review]
Thanks @CrazyParanoid. I've left some feedback inline.

In addition to that, would you please add documentation based on the existing servlet docs? If you don't have time to do that, can you please create a ticket?

File: config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java (hunk 1 above)
Comment (CrazyParanoid) [reply config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java]: I added initialization of `oneTimeTokenService` with `InMemoryReactiveOneTimeTokenService`, but I left the `getOneTimeTokenService` method in order to be able to get `oneTimeTokenService` as a bean.

File: config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java [hunk 4]
```diff
@@ -244,7 +253,7 @@
  *     }
  * }
  * </pre>
- *
+ * <p>
```
[the remark this answers is not in the document]
Comment (CrazyParanoid) [reply config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java]: Sorry, fixed.

**Comment by CrazyParanoid** [comment]
Hi @marcusdacoregio. Thanks for your feedback! All your comments have been resolved. 
I'll add documentation for `Reactive One-Time Token Login` in the next couple of days.

**Comment by CrazyParanoid** [comment]
Hi @marcusdacoregio ! I have resolved all errors and conflicts and added documentation. Could you please review it?

**Review by jzheaux — CHANGES_REQUESTED** [review]
Hey, @CrazyParanoid! Thanks for the updates; good progress.

I've left some feedback inline. Also, I've created a sample at https://github.com/jzheaux/spring-security-samples/tree/gh-15699/reactive/webflux/java/authentication/one-time-token/magic-link that you use to test the full roundtrip. You can use the `--include-build` Gradle feature to include your local Spring Security as part of the build for the sample.

File: web/src/main/java/org/springframework/security/web/server/authentication/ott/GenerateOneTimeTokenWebFilter.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+			.switchIfEmpty(chain.filter(exchange).then(Mono.empty()))
+			.flatMap((matchResult) -> resolveUsername(exchange.getRequest()))
+			.switchIfEmpty(chain.filter(exchange).then(Mono.empty()))
+			.map(GenerateOneTimeTokenRequest::new)
+			.flatMap(this.oneTimeTokenService::generate)
+			.flatMap((token) -> this.generatedOneTimeTokenHandler.handle(exchange, token));
+	}
+
+	private Mono<String> resolveUsername(ServerHttpRequest request) {
+		String username = request.getQueryParams().getFirst("username");
```
[the remark this answers is not in the document]
Comment (CrazyParanoid) [reply web/src/main/java/org/springframework/security/web/server/authentication/ott/GenerateOneTimeTokenWebFilter.java:69]: I can suggest adding the `allowFormEncodedBodyParameter` property. In the servlet part, the logic for converting the request into h is built on the query parameter “token”. It will be strange if the servlet and reactive parts are very different, especially since the servlet part is already in the main branch.

File: web/src/main/java/org/springframework/security/web/server/authentication/ott/ServerOneTimeTokenAuthenticationConverter.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+	public Mono<Authentication> convert(ServerWebExchange exchange) {
+		String token = resolveTokenFromRequest(exchange.getRequest());
+		if (!StringUtils.hasText(token)) {
+			return Mono.empty();
+		}
+		return Mono.just(OneTimeTokenAuthenticationToken.unauthenticated(token));
+	}
+
+	private String resolveTokenFromRequest(ServerHttpRequest request) {
+		List<String> parameterTokens = request.getQueryParams().get("token");
```
[the remark this answers is not in the document]
Comment (CrazyParanoid) [reply web/src/main/java/org/springframework/security/web/server/authentication/ott/ServerOneTimeTokenAuthenticationConverter.java:51]: I can suggest adding the allowFormEncodedBodyParameter property. In the servlet part, the logic for converting the request into h is built on the query parameter “token”. It will be strange if the servlet and reactive parts are very different, especially since the servlet part is already in the main branch.

File: core/src/test/java/org/springframework/security/authentication/ott/reactive/InMemoryReactiveOneTimeTokenServiceTests.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+				.block();
+			generated.add(oneTimeToken);
+		}
+		return generated;
+	}
+
+	private void setClock(Clock clock) {
+		OneTimeTokenService innerOneTimeTokenService = (OneTimeTokenService) ReflectionTestUtils
+			.getField(this.oneTimeTokenService, "oneTimeTokenService");
+		ReflectionTestUtils.setField(innerOneTimeTokenService, "clock", clock);
```
[the remark this answers is not in the document]
Comment (CrazyParanoid) [reply core/src/test/java/org/springframework/security/authentication/ott/reactive/InMemoryReactiveOneTimeTokenServiceTests.java:137]: Method `setClock` already exists, but for some reason it is package private. I opened [gh-15863](https://github.com/spring-projects/spring-security/issues/15863)

**Comment by CrazyParanoid** [comment]
Hi @jzheaux, thanks for feedback. Special thanks for the example, I tested with local version `6.4.0-SNAPSHOT`. It works fine. I also tested several additional cases. Could you please review my changes?

File: web/src/main/java/org/springframework/security/web/server/ui/OneTimeTokenSubmitPageGeneratingWebFilter.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+			.getBytes(Charset.defaultCharset());
+	}
+
+	private static String csrfToken(CsrfToken token) {
+		return HtmlTemplates.fromTemplate(CSRF_INPUT_TEMPLATE)
+			.withValue("name", token.getParameterName())
+			.withValue("value", token.getToken())
+			.render();
+	}
+
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/server/ui/OneTimeTokenSubmitPageGeneratingWebFilter.java:100]: Please add JavaDoc. Something like

```java
/** 
  * Use this {@link RequestMatcher} to choose whether this filter will
  * handle the request. By default, it handles {@code /ott/submit}.
  */
```

File: web/src/main/java/org/springframework/security/web/server/ui/OneTimeTokenSubmitPageGeneratingWebFilter.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+	public void setRequestMatcher(ServerWebExchangeMatcher requestMatcher) {
+		Assert.notNull(requestMatcher, "requestMatcher cannot be null");
+		this.matcher = requestMatcher;
+	}
+
+	public void setLoginProcessingUrl(String loginProcessingUrl) {
```
Comment (jzheaux) [inline web/src/main/java/org/springframework/security/web/server/ui/OneTimeTokenSubmitPageGeneratingWebFilter.java:106]: Please add JavaDoc

File: config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java [hunk 10]
```diff
@@ -1640,6 +1687,9 @@ else if (this.securityContextRepository != null) {
 		if (this.logout != null) {
 			this.logout.configure(this);
 		}
+		if (this.oneTimeTokenLogin != null) {
+			this.oneTimeTokenLogin.configure(this);
```
Comment (jzheaux) [inline config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java:1691]: This will need to include negotiating the security context repository (see the way that `formLogin` and `oauth2Login` are configured in this method). To achieve this, `OneTimeTokenLoginSpec` needs a `securityContextRepository` method.

Because `AuthenticationWebFilter` defaults to a no-op security context repository, I believe this contributes to why the user is not ultimately logged in after posting the token.

File: config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java [hunk 11]
```diff
@@ -2343,8 +2392,7 @@ public CsrfSpec accessDeniedHandler(ServerAccessDeniedHandler accessDeniedHandle
 
 		/**
 		 * Configures the {@link ServerCsrfTokenRepository} used to persist the CSRF
-		 * Token. Default is
-		 * {@link org.springframework.security.web.server.csrf.WebSessionServerCsrfTokenRepository}.
+		 * Token. Default is {@link WebSessionServerCsrfTokenRepository}.
```
Comment (jzheaux) [inline config/src/main/java/org/springframework/security/config/web/server/ServerHttpSecurity.java:2395]: Please leave unrelated changes out of the PR. (It looks like a few others in this file sneaked in)

**Review by jzheaux — CHANGES_REQUESTED** [review]
Thanks for the quick turnaround, @CrazyParanoid, this is coming together. I've left another round of feedback that is based on your recent changes.

Also, the sample does not yet work for me with this code. When I visit the URL in the email, it correctly posts the token, but it does not result in the user being logged in. Please see my inline comments for more details on one item that is contributing to this.

File: core/src/test/java/org/springframework/security/authentication/ott/reactive/InMemoryReactiveOneTimeTokenServiceTests.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (CrazyParanoid) [reply core/src/test/java/org/springframework/security/authentication/ott/reactive/InMemoryReactiveOneTimeTokenServiceTests.java]: It is better to change the `OneTimeTokenService` type to `InMemoryOneTimeTokenService` for inner `oneTimeTokenService` and add the `setClock` method.

File: web/src/main/java/org/springframework/security/web/server/authentication/ott/GenerateOneTimeTokenWebFilter.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+	public GenerateOneTimeTokenWebFilter(ReactiveOneTimeTokenService oneTimeTokenService,
+			ServerGeneratedOneTimeTokenHandler generatedOneTimeTokenHandler) {
+		Assert.notNull(generatedOneTimeTokenHandler, "generatedOneTimeTokenHandler cannot be null");
+		Assert.notNull(oneTimeTokenService, "oneTimeTokenService cannot be null");
+		this.generatedOneTimeTokenHandler = generatedOneTimeTokenHandler;
+		this.oneTimeTokenService = oneTimeTokenService;
+	}
+
+	@Override
+	public Mono<Void> filter(ServerWebExchange exchange, WebFilterChain chain) {
```
Comment (CrazyParanoid) [inline web/src/main/java/org/springframework/security/web/server/authentication/ott/GenerateOneTimeTokenWebFilter.java:57]: This code produced `ServerHttpResponse already committed (200 OK)`. It's fixed now.

File: web/src/main/java/org/springframework/security/web/server/authentication/ott/GenerateOneTimeTokenWebFilter.java [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+		// @formatter:off
+		return this.matcher.matches(exchange)
```
Comment (CrazyParanoid) [inline web/src/main/java/org/springframework/security/web/server/authentication/ott/GenerateOneTimeTokenWebFilter.java:59]: This code produced `ServerHttpResponse already committed`. It's fixed now.

**Review by jzheaux — APPROVED** [review]

**Comment by jzheaux** [comment]
Thanks for all your hard work on this, @CrazyParanoid. It's now merged to `main`.
